### PR TITLE
dma: Scatter gather test

### PR DESCRIPTION
**Pull request by teburd**
Adds a scatter gather test against memory to memory transfers.

Based on #43984 so the updated designware driver can be built on both test cases in #43883 

DesignWare DMA passes both tests, NXP's eDMA does not seem to pass the scatter gather test and its not clear as to why, maybe @hakehuang wouldn't mind taking a look?

Marked DNM until #43984 is merged

### CODE REVIEW

**Comment by danieldegrasse** [comment]
@hakehuang the changes to `fsl_edma` in MCUXpresso SDK 2.11 (#43826) and the updates to the EDMA driver in #42750 enable this test to pass on an RT1060.

@dleach02 @DerekSnell FYI- This looks like the same issues we've been seeing with the EDMA driver.

**Comment by teburd** [comment]
> @hakehuang the changes to `fsl_edma` in MCUXpresso SDK 2.11 (#43826) and the updates to the EDMA driver in #42750 enable this test to pass on an RT1060.
> 
> @dleach02 @DerekSnell FYI- These look like the same issues we've been seeing with the EDMA driver, although I didn't investigate too closely (just tried updating the HAL and our EDMA driver)

Thanks for taking a look at this and confirming that there are fixes required for this to work on eDMA. I've pushed a change to this PR to disable it for all platforms initially, in the updated designware DMA PR I've enabled it for the appropriate intel adsp platforms. When the NXP changes are merged we can enable it there as well.

**Comment by hakehuang** [comment]
> @hakehuang Did you pull in the updates to the EDMA driver in #42750? I saw errors like that when I only had updated the HAL.

Thanks @danieldegrasse , there is one key update at 

```
#ifdef CONFIG_HAS_MCUX_CACHE

#ifdef CONFIG_DMA_MCUX_USE_DTCM_FOR_DMA_DESCRIPTORS

#if DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_dtcm), okay)
#define EDMA_TCDPOOL_CACHE_ATTR __dtcm_noinit_section
#else /* DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_dtcm), okay) */
#error Selected DTCM for MCUX DMA descriptors but no DTCM section.
#endif /* DT_NODE_HAS_STATUS(DT_CHOSEN(zephyr_dtcm), okay) */

#elif defined(CONFIG_NOCACHE_MEMORY)
#define EDMA_TCDPOOL_CACHE_ATTR __nocache
#else
#error tcdpool could not be located in cacheable memory, a requirement for proper EDMA operation.
#endif /* CONFIG_DMA_MCUX_USE_DTCM_FOR_DMA_DESCRIPTORS */

#else /* CONFIG_HAS_MCUX_CACHE */

#define EDMA_TCDPOOL_CACHE_ATTR

#endif /* CONFIG_HAS_MCUX_CACHE */

static __aligned(32) EDMA_TCDPOOL_CACHE_ATTR edma_tcd_t
tcdpool[DT_INST_PROP(0, dma_channels)][CONFIG_DMA_TCD_QUEUE_SIZE];
```
which fixes my issues herre

**Review by hakehuang — DISMISSED** [review]
need wait for #42750 merged. others looks OK for me and test pass on my rt series board

**Review by danieldegrasse — DISMISSED** [review]

**Comment by laurenmurphyx64** [comment]
I don't think it matters a whole lot, especially considering you just got 2 approvals on this and updating the PR would dismiss them and put you back in review purgatory, but the copyrights for the new files should be 2022 instead of 2021. You can always fix the copyright comments in another little PR after this one gets merged - it'd get merged super quick since it wouldn't even touch code.

File: tests/drivers/dma/scatter_gather/src/test_dma_sg.c
Comment (laurenmurphyx64) [inline tests/drivers/dma/scatter_gather/src/test_dma_sg.c:13]:
```suggestion
 *   -# Set direction memory-to-memory
```
Just a nitpick. We have a bunch of little things like this in the codebase already, so it's not a huge deal.

Comment (teburd) [reply tests/drivers/dma/scatter_gather/src/test_dma_sg.c:13]: I wish I could change this and not dismiss the reviews, a downside of the new setup

Comment (danieldegrasse) [reply tests/drivers/dma/scatter_gather/src/test_dma_sg.c:13]: I'll reapprove if you push now- assuming @laurenmurphyx64 can provide second approval

Comment (laurenmurphyx64) [reply tests/drivers/dma/scatter_gather/src/test_dma_sg.c:13]: @danieldegrasse Yep :-)

**Review by danieldegrasse — DISMISSED** [review]

**Review by laurenmurphyx64 — DISMISSED** [review]

**Review by hakehuang — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

File: tests/drivers/dma/scatter_gather/src/test_dma_sg.c
Comment (ceolin) [inline tests/drivers/dma/scatter_gather/src/test_dma_sg.c:87]: Is this conditional really necessary ? I don't see the callback using the data passsed here. The driver should not do anything with the user pointer as well.

I may have missed something though.

File: tests/drivers/dma/scatter_gather/src/test_dma_sg.c
Comment (ceolin) [inline tests/drivers/dma/scatter_gather/src/test_dma_sg.c:109]: Can't these address be 64bit if `CONFIG_DMA_64BIT` is selected ? Btw, why don't we use `uintptr_t` for these address instead of having it conditionally set ? (not related with this particular patch, just a question)

**Review by ceolin — APPROVED** [review]
Minor comments, looks good to me.

**Pull request merged**
